lib/phoenix_live_view.ex

```diff
@@ -473,32 +473,65 @@ defmodule Phoenix.LiveView do
       require Phoenix.LiveView.Renderer
       @before_compile Phoenix.LiveView.Renderer
 
-      @doc false
-      def __live__, do: unquote(Macro.escape(Phoenix.LiveView.__live__(__MODULE__, opts)))
+      @phoenix_live_opts opts
+      Module.register_attribute(__MODULE__, :phoenix_live_on_mount, accumulate: true)
+      @before_compile Phoenix.LiveView
     end
   end
 
-  @doc false
-  def __live__(module, opts) do
-    container = opts[:container] || {:div, []}
-    namespace = opts[:namespace] || module |> Module.split() |> Enum.take(1) |> Module.concat()
-    name = module |> Atom.to_string() |> String.replace_prefix("#{namespace}.", "")
+  defmacro __before_compile__(_env) do
+    quote do
+      @doc false
+      def __live__ do
+        alias Phoenix.LiveView.Lifecycle
+
+        module = __MODULE__
+        opts = @phoenix_live_opts
+
+        container = opts[:container] || {:div, []}
+
+        namespace =
+          opts[:namespace] || module |> Module.split() |> Enum.take(1) |> Module.concat()
+
+        name = module |> Atom.to_string() |> String.replace_prefix("#{namespace}.", "")
 
-    layout =
-      case opts[:layout] do
-        {mod, template} when is_atom(mod) and is_binary(template) ->
-          {mod, template}
+        layout =
+          case opts[:layout] do
+            {mod, template} when is_atom(mod) and is_binary(template) ->
+              {mod, template}
 
-        nil ->
-          nil
+            nil ->
+              nil
 
-        other ->
-          raise ArgumentError,
-                ":layout expects a tuple of the form {MyLayoutView, \"my_template.html\"}, " <>
-                  "got: #{inspect(other)}"
+            other ->
+              raise ArgumentError,
+                    ":layout expects a tuple of the form {MyLayoutView, \"my_template.html\"}, " <>
+                      "got: #{inspect(other)}"
+          end
+
+        lifecycle = Lifecycle.mount(module, @phoenix_live_on_mount)
+
+        %{
+          container: container,
+          name: name,
+          kind: :view,
+          module: module,
+          layout: layout,
+          lifecycle: lifecycle
+        }
       end
+    end
+  end
 
-    %{container: container, name: name, kind: :view, module: module, layout: layout}
+  @doc """
+  Declares a hook to be attached to the `:mount` lifecycle event.
+
+  For more information about lifecycle hooks, see `attach_hook/4`.
+  """
+  defmacro on_mount(id) do
+    quote do
+      Module.put_attribute(__MODULE__, :phoenix_live_on_mount, unquote(id))
```

Comment:
It would be nice to validate it here, so the error message is as close as possible to the source:

    Module.put_attribute(__MODULE__, :phoenix_live_on_mount, Phoenix.LiveView.Lifecycle.on_mount(unquote(id)))